Add PIPECG reference implementation

gojakuch opened the pull request:
This PR adds the reference implementation for the pipelined CG method.

gojakuch commented:
I've put the "Ready for review" label, but I'm not sure if I should've put the "Requires feedback" one instead, since this is my first PR here and it's hard for me to assess the state of it. any feedback would be considered of course :)

yhmtsai commented on core/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,301 @@
+    r->compute_conj_dot(z, rho, reduction_tmp);
+    // delta = dot(w, z)
+    w->compute_conj_dot(z, delta, reduction_tmp);
+    // can these two dot products above be merged in some way?
```
```suggestion
    // TODO: can these two dot products above be merged in some way?
```
the current support should allow [r w] and [z z], so you need to duplicate the z unfortunately.
It might be possible if we introduce conjugate transposed apply on dense it will be like [r w]' * z

pratikvn replied:
Yes, that would be the plan, but we will do that in a later PR.

gojakuch replied:
@pratikvn so should I remove this comment for now? I've put it there so that we don't forget about it

yhmtsai replied:
I would say we can keep that for reminder

pratikvn reviewed (changes requested):
Nice work! Mostly looks good to me. Only the documentation needs to be updated.

gojakuch commented on reference/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,164 @@
+            if (stop_status->get_const_data()[j].has_stopped()) {
+                continue;
+            }
+            if (is_nonzero(beta->at(j))) {
```
yes, good point, that's a problem I haven't noticed. I just did what's done in the CG in this case, which wouldn't work out for the PIPECG logic. If beta becomes zero somewhere throughout the iteration, for PIPECG, it is gonna stay 0 forever and no other values are going to be changed. I think one way to address this would be to check if beta is zeroed out by step_2 and prevent this from happening there, possibly by just letting `beta = delta`, which is a valid value, as it's what we initialise beta with and what we set beta to in step_1 in its corner case.

pratikvn replied:
Yes, I think that is a reasonable choice.

gojakuch commented on include/ginkgo/core/solver/pipe_cg.hpp:
```diff
@@ -0,0 +1,184 @@
+/**
+ * PIPE_CG or the pipelined conjugate gradient method is an iterative type
+ * Krylov subspace method which is suitable for symmetric positive definite
+ * methods. It improves upon the CG method by allowing computation of inner
+ * products and norms to be overlapped with operator and preconditioner
+ * application. The pipelined method scales up to the 10^6 nodes of the assumed
+ * exascale machine, while its standard counterpart level off about one order of
+ * magnitude earlier, as suggested in the referenced paper (see below).
```
added

pratikvn reviewed (changes requested):
Great! Thanks for making the changes. One last thing before we merge it is that you add yourself to the contributors list here after Marcel Koch: https://github.com/ginkgo-project/ginkgo/blob/develop/contributors.txt#L20 with the following info. `LastName FirstName <email> Affiliation`

pratikvn reviewed (approved):
LGTM! Thank you for your contribution!

yhmtsai commented on core/test/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,272 @@
+#include "core/test/utils.hpp"
+
+
+namespace {
```
```suggestion
```
we do not need the anonymous namespace for test now

yhmtsai commented on core/test/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,272 @@
+}
+
+
+}  // namespace
```
same here

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+        this->small_n.get(), this->small_prev_rho.get(), this->small_rho.get(),
+        this->small_delta.get(), &this->small_stop);
+
+    GKO_ASSERT_MTX_NEAR(this->small_beta, this->small_delta, 0);
```
also need to test the other vector

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+
+TYPED_TEST(PipeCg, KernelStep1)
+{
+    this->small_x->fill(1);
+    this->small_r->fill(2);
+    this->small_z->fill(3);
+    this->small_w->fill(4);
+
+    this->small_p->fill(4);
+    this->small_q->fill(3);
+    this->small_f->fill(2);
+    this->small_g->fill(1);
+
+    this->small_rho->at(0) = 2;
+    this->small_rho->at(1) = 3;
+    this->small_beta->at(0) = 8;
+    this->small_beta->at(1) = 3;
+    this->small_stop.get_data()[1] = this->stopped;
```
follow AAA rules, but you can use comment to distinguish the group

yhmtsai replied:
same for the rest of test in this file

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+                      gko::stop::ResidualNorm<value_type>::build()
+                          .with_reduction_factor(r<value_type>::value))
+                  .on(exec)),
+          mtx_big(gko::initialize<Mtx>(
```
the actual initiation follows the member declaration.
here you need to make it to follow the same order of member.

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+
+    solver->apply(b, x);
+
+    GKO_ASSERT_MTX_NEAR(x, l({81.0, 55.0, 45.0, 5.0, 85.0, -10.0}),
```
it is 50 times the CG test

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+    solver->apply(b, x);
+
+    GKO_ASSERT_MTX_NEAR(x, l({33.0, -56.0, 81.0, -30.0, 21.0, 40.0}),
+                        r<value_type>::value * 2 * 1e5);
```
it is 2000 times of CG test. it sounds like we need to check something again.

gojakuch replied:
this is a weird thing that I don't get. the numbers that my PIPECG implementation outputs here are the same as the tested values, and yet it fails this check. I explained this to Pratik and he fixed it by adjusting these tolerances. I've got no other idea what that could be. the output numbers are literally the same here. any suggestions?

yhmtsai replied:
do you mean x is [33.0, -56.0, 81.0, -30.0, 21.0, 40.0]' on output?

gojakuch replied:
yeah. or do I not get what these tolerances are? :sweat_smile:

gojakuch replied:
yeah nevermind, I misinterpreted some things in my test example, it's out of the tolerance bounds

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,694 @@
+}
+
+
+}  // namespace
```
same here

yhmtsai commented on core/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,301 @@
+    const config::type_descriptor& td_for_child)
+{
+    auto params = solver::PipeCg<ValueType>::build();
+    common_solver_parse(params, config, context, td_for_child);
```
need to add corresponding test in config core/test/config/solver.cpp and add it into the main parse
core/config/solver_config.cpp and core/config/registry.cpp

gojakuch replied:
alright, I'll add these tests

yhmtsai commented on core/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,301 @@
+                                         &stop_status));
+    // r = r - Ax
+    this->get_system_matrix()->apply(neg_one_op, dense_x, one_op, r);
+    // z = preconditioner * r
```
if there is no specific reason, I will suggest the variable follows the paper notation.

gojakuch replied:
if you're talking about variable naming, Ginkgo I chose names compatible with the CG implementation in Ginkgo. the naming in the paper contradicts the naming in CG in Ginkgo

yhmtsai commented on core/solver/pipe_cg.cpp:
```diff
@@ -0,0 +1,301 @@
+        // delta = dot(w, z)
+        w->compute_conj_dot(z, delta, reduction_tmp);
+        // check
+        ++iter;
```
I think the criterion check should be earlier?

gojakuch replied:
so are you suggesting that I put the check before computing rho in the loop?

yhmtsai replied:
one additional check after the first rho computation

yhmtsai reviewed (changes requested):
need to add small section into test/test_install/test_install.cpp

pratikvn commented:
It seems nvhpc compilers still have some issues with the pipecg tolerances, so it might need to be increased even more: https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/9693465904

gojakuch commented:
> It seems nvhpc compilers still have some issues with the pipecg tolerances, so it might need to be increased even more: https://gitlab.com/ginkgo-project/ginkgo-public-ci/-/jobs/9693465904

I think @yhmtsai suggests that we find another solution

pratikvn commented:
The tolerance problem is not limited to PipeCG, but is prevalent across **all** solvers in Ginkgo. We have arbitrarily increased the tolerances for different solvers to ensure that they pass the tests. The matrices we use to test the solvers are also for the most part arbitrary (we have some control: hpd-ness, sparsity, but not over condition number for example.) 

A proper solution IMO would be to have full control over the matrix and vector inputs: 
1. Generate matrices with specific condition numbers and eigenvalue distributions.
2. Define vectors with controlled vector norms.

Then we can correlate solvers with inputs and tailor the inputs rather than tailor the output tolerances. This approach requires a more sophisticated sparse matrix generation, which we currently do not have.

Therefore, IMO, the current approach should be okay, and we should revisit the controlled inputs approach at a later point (with a new PR) and update all the solvers and their tests.

gojakuch commented:
@yhmtsai please consider @pratikvn's comment when reviewing. I rebased this and tried to address all the other things you've mentioned except for the tolerances

yhmtsai commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,678 @@
+
+    solver->apply(b, x);
+
+    GKO_ASSERT_MTX_NEAR(x, l({33.0, -56.0, 81.0, -30.0, 21.0, 40.0}),
```
I suggest to add a TODO here.

yhmtsai replied:
I think the sentence until ...design better tests should be good enough

yhmtsai reviewed (changes requested):
I do not fully agree Pratik's comments.
Pratik's comments should only for the general testing case for different executor.
In the reference test, everything is initialized manually, so everything is under control.
I have checked the implementation with the paper and I do not find any difference between them. Because the paper also showed less convergence than CG, I do not hold this PR due to the relaxed error check on big system.

The change request is for the rest of my comments.

yhmtsai commented:
still miss the first check when having the first rho

gojakuch commented:
> still miss the first check when having the first rho and the test_install

sorry, I don't think I understood what you mean here

gojakuch commented on reference/test/solver/pipe_cg_kernels.cpp:
```diff
@@ -0,0 +1,678 @@
+    solver->apply(b, x);
+
+    GKO_ASSERT_MTX_NEAR(x, l({81.0, 55.0, 45.0, 5.0, 85.0, -10.0}),
+                        r<value_type>::value * 5 * 1e3);
```
ok, so what should I put as the tolerance then?

gojakuch commented:
rebased the PR because the formatting bot complained about something unrelated. seems to have fixed that

yhmtsai reviewed (changes requested):
I think it still misses the part in ` test/test_install/test_install.cpp`.
It needs a section like
```
// core/solver/pipe_cg.hpp
{
    using Solver = gko::solver::PipeCg<>;
    check_solver<Solver>(exec, A_raw, b, x);
}
```
I also add the place which should contains the check.
you can use only one check but you need to rearrange the algorithm like CG
Otherwise, use one out of the loop and another in the loop after you get the updated residual norm.

gojakuch commented:
> I think it still misses the part in ` test/test_install/test_install.cpp`. It needs a section like
> 
> ```
> // core/solver/pipe_cg.hpp
> {
>     using Solver = gko::solver::PipeCg<>;
>     check_solver<Solver>(exec, A_raw, b, x);
> }
> ```
As I've mentioned in my comment before: I think another check is failing now, I suspect that it's because of me adding PipeCG to test_install like Mike suggested, while the unified kernels are not implemented in this PR. I think I'll just add it to the test_install in my next PR with kernel implementation

meaning that I've tried adding exactly this block of code but a new check started failing after this. I'll try this again now, but I can just do this in my next PR where I implement the kernels.

as for the check, I don't see how adding two checks is better, but I'll add this.

yhmtsai reviewed (approved):
LGTM. sorry for missing your comment and confusing suggestion.
Yes, test_install indeed also tests for different backend, so it will be failed when the other backend are enabled.
Only one thing left is the iteration count needs to be moved a little.
Thanks for implementing it!

gojakuch commented:
> LGTM. sorry for missing your comment and confusing suggestion. Yes, test_install indeed also tests for different backend, so it will be failed when the other backend are enabled. Only one thing left is the iteration count needs to be moved a little. Thanks for implementing it!

@yhmtsai thanks! I've rebased the PR and moved the iter counter before the second check, so that we don't check with iter==0 twice

pratikvn commented:
Thank you @gojakuch for your contribution!
